### ISSUE TITLE
"Can't resolve '@xxx' in " Yarn workspace

### ISSUE THREAD

**Issue by fabb** [martpie/next-transpile-modules#198]
- [x] I HAVE READ [THE FAQ](https://github.com/martpie/next-transpile-modules#faq) AND MY PROBLEM WAS NOT DESCRIBED THERE
- [ ] I WILL GIVE 10$ TO CHARITY IF MY SOLUTION WAS ACTUALLY IN THE README/FAQ

**Are you trying to transpile a local package or an npm package?**

A local package which is a yarn workspace

**Describe the bug**
A clear and concise description of what the bug is (don't forget the logs if applicable).

I'm upgrading from next-transpile-modules v4 and next.js with webpack 4 to next.js 10.2 with webpack 5 and next-transpile-modules v7. I have a directory structure as follows:

* applications
    * app1
        * node_modules
        * package.json
        * next.config.js
* common
    * package.json
* node_modules
    * @wh
        * app1
        * common
* package.json

The root package.json references `applications/app1` and `common` as workspaces which use the prefix "@wh". Yarn creates links to all workspaces in the root node_modules under the prefix `@wh`.

next-transpile-modules is integrated in next.config.js like this: `require('next-transpile-modules')(['@wh'])`.

When starting the application with `next dev`, this error is shown:
```
getPackageRootDirectory Error: Can't resolve '@wh' in '<my_root>/applications/app1'.
```

I added some console.logs to next-transpile-modules and it looks like the `@wh` package is only searched in `applications/app1/node_modules`, but it's not there since yarn workspaces only puts it into the root `node_modules`.

**To Reproduce**
Steps to reproduce the behavior (a reproduction is even better):

1. Create a root package.json and several workspaces with yarn (configure the `name` of the packages to have a common prefix, e.g. `@wh/app1`)
3. In one of the workspaces, setup next.js with next-transpile-modules.
4. Configure next-transpile-modules to use the workspace package prefix

**Expected behavior**

I would expect that next-transpile-modules would recognize that it runs in a yarn workspace, and tries to look up the packages to be transpile in the root package folder. Alternatively, it should be possible to set the root folder explicitly in the next-transpile-modules config (I tried to use `require('next-transpile-modules')(['../../@wh'])`, but that won't work either).

**Setup**

- Next.js version: 10.2
- `next-transpile-modules` version: 7.0.0
- Node.js version: 14.16.1
- `npm`/`yarn` version: 1.22.10
- Operating System: macOS Big Sur
- Webpack 4 or 5: 5

**Additional context**

It used to work in next-transpile-modules v4.

**Comment by fabb** [comment]
Sorry, it was a long day. In fact I did carefully read through the upgrade notes and skimmed through the readme before creating this issue. If the answer is in there, I don‘t understand it, and I didn‘t want to be „punished“ for not understanding something (don‘t mistake this with not wanting to give to charity, I do in other circumstances).

I was especially confused by the part „Ensure all your transpiled modules have a main field. If you think it does not make sense, still, have one!“ since the linked next example does not have a `main` field: https://github.com/vercel/next.js/blob/canary/examples/with-yarn-workspaces/packages/web-app/package.json
Anyways, I had it there, and also removing did not make a difference. I debugged through the code of this lib, and initially `CWD` is set to the `applications/app1` Folder, and the call to `resolve` throws. I didn‘t yet find out why and how the code is supposed to detect the root package of a yarn workspace.

**Comment by martpie** [comment]
The example in the Next.js is using ntm v4, so this is normal. I encourage you to have a look at https://github.com/martpie/monorepo-typescript-next-the-sane-way

**Comment by martpie** [comment]
> I had it there

It's not only about having a main field, it's about pointing it to a real JavaScript module

**Comment by fabb** [comment]
> The example in the Next.js is using ntm v4, so this is normal. I encourage you to have a look at https://github.com/martpie/monorepo-typescript-next-the-sane-way

Ah thanks. I notice in this example that it does not use a prefix in the `shared` package module, maybe that‘s the issue? My monorepo package names all are prefixed with `@wh/`. Could it be that this is no longer supported by next-transpile modules 5+?

> > I had it there
> 
> It's not only about having a main field, it's about pointing it to a real JavaScript module

I tried creating an empty `index.js` as suggested in the changelog, did not solve the issue.

**Comment by fabb** [comment]
Ok, it really looks like it is a bug/limitation of next-transpile-modules with scoped packages. I modified [monorepo-typescript-next-the-sane-way](https://github.com/martpie/monorepo-typescript-next-the-sane-way) to use package scopes, and then I run into the same error:
```
Error: next-transpile-modules - an unexpected error happened when trying to resolve "shared"
Error: Can't resolve 'shared' in '/some-path/monorepo-typescript-next-the-sane-way/website'
```

Here's a branch that can be checked out for reproduction: https://github.com/martpie/monorepo-typescript-next-the-sane-way/pull/19

Is it a bug or a limitation that's out of scope (pun intended 😅)?

**Comment by martpie** [comment]
> Is it a bug or a limitation that's out of scope (pun intended 😅)?

Pretty sure this is a problem with your setup ;) a lot of people use ntm and scoped packages. I replied on your PR

**Comment by fabb** [comment]
It indeed was. I just found it out, and feel very stupid. Looks like a good night of sleep helps wonders. Yesterday I tried replacing `require('next-transpile-modules')(['@wh'])` with something like `require('next-transpile-modules')(['@wh/common', '@wh/applications/app1'])` to list all the `@wh` packages. But of course this makes no sense because I just need to transpile `@wh/common` in the context of app1 🙈.
Thank you for helping.

**Comment by martpie** [comment]
Oh, I missed that: you cannot `require('next-transpile-modules')(['@wh'])` because it is not a module.

But you can `require('next-transpile-modules')(['@wh/common'])`.

See https://github.com/martpie/next-transpile-modules/issues/143

**Comment by fabb** [comment]
I have another question also related to the "Can't resolve" issue somehow. I had problems running my next application in my docker container, I got the "Can't resolve" error again while it worked fine in dev mode. Note that I'm not using `next start`, but use a custom express server.

I found out that in my docker build I also need to copy the package.json + index.js from `@wh/common` to the final application folder since it looks like next-transpile-modules is also running there. Should it really run for the final application? There is nothing to transpile anymore, the built next application in the `.next` folder does not contain any `@wh` imports anymore, only the final webpack chunks.
